registration/static/js/src/registration.js

```diff
@@ -54,10 +53,45 @@ app.controller('Registration', ['$scope', '$http', 'djangoForm', function($scope
         return getFormFields().filter(key => $scope.form[key].$dirty || $scope.form[key].$message);
     };
 
+    // Returns the result of validating password using zxcvbn.
+    var validatePassword = function() {
+        var password = $scope.form.password.$viewValue;
+        return zxcvbn(password);
+    };
+
+    // Makes sure each suggestion ends with period.
+    var formatSuggestions = function(suggestions) {
+        return _.map(suggestions, function(suggestion) {
+            if (!suggestion.endsWith('.')) {
+                return suggestion + '.';
+            }
+            return suggestion;
+        });
+    };
+
+    // Returns feedback from password validation.
+    var getValidationFeedback = function() {
+        var validationResult = validatePassword(),
+            suggestions = validationResult.feedback.suggestions,
+            warning = validationResult.feedback.warning;
+        if (warning) {
+            suggestions.push(warning);
+        }
+        suggestions = formatSuggestions(suggestions);
+        // Concatenate suggestions to ensure that all of them are displayed
+        return [suggestions.join(" ")];
+    };
+
     // Display the given error messages, making sure not to clear messages
     // that are already displayed.
     var displayErrors = function(errors) {
         errors = _.pick(errors, getModifiedFields());
+        if (errors.password && $scope.form.password.$viewValue) {
```

Comment:
@omarkhan Thanks for the pointers! Being able to avoid a roundtrip to the server is great. I updated the code, keeping the default error message from the server as a fallback for the odd case where `zxcvbn` doesn't produce any suggestions or warnings for a given password.

However, I'm now having some issues validating that the client calculated password strength: This now has to be done when the user submits the form, so the `password_strength` field has to be set to `zxcvbn(password).score` on the client. Do you know how to do that (in a way where the value actually reaches the server when submitting the form)? I found that I could set `$scope.form.password_strength.$viewValue` successfully on the client, but the server still didn't receive the value that I chose for it. Same result when setting `$modelValue` or using `$setViewValue(score)`.